fvm/environment/value_store.go

```diff
@@ -0,0 +1,110 @@
+package environment
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/trace"
+)
+
+// ValueStore provides read/write access to the account storage.
+type ValueStore struct {
+	tracer *Tracer
+	meter  Meter
+
+	accounts Accounts
+}
+
+func NewValueStore(tracer *Tracer, meter Meter, accounts Accounts) *ValueStore {
+	return &ValueStore{
+		tracer:   tracer,
+		meter:    meter,
+		accounts: accounts,
+	}
+}
+
+func (store *ValueStore) GetValue(owner, key []byte) ([]byte, error) {
+	var valueByteSize int
+	span := store.tracer.StartSpanFromRoot(trace.FVMEnvGetValue)
+	defer func() {
+		if !trace.IsSampled(span) {
+			span.SetAttributes(
+				attribute.String("owner", hex.EncodeToString(owner)),
+				attribute.String("key", string(key)),
+				attribute.Int("valueByteSize", valueByteSize),
+			)
+		}
+		span.End()
+	}()
+
+	v, err := store.accounts.GetValue(
+		flow.BytesToAddress(owner),
+		string(key),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("get value failed: %w", err)
+	}
+	valueByteSize = len(v)
+
+	err = store.meter.MeterComputation(
+		ComputationKindGetValue,
+		uint(valueByteSize))
+	if err != nil {
+		return nil, fmt.Errorf("get value failed: %w", err)
+	}
+	return v, nil
+}
+
+// TODO disable SetValue for scripts, right now the view changes are discarded
+func (store *ValueStore) SetValue(owner, key, value []byte) error {
+	span := store.tracer.StartSpanFromRoot(trace.FVMEnvSetValue)
+	if !trace.IsSampled(span) {
+		span.SetAttributes(
+			attribute.String("owner", hex.EncodeToString(owner)),
+			attribute.String("key", string(key)),
+		)
+	}
+	defer span.End()
+
+	err := store.meter.MeterComputation(
+		ComputationKindSetValue,
+		uint(len(value)))
+	if err != nil {
+		return fmt.Errorf("set value failed: %w", err)
```

Comment:
nit: i know it is not the change from this PR - the msg here can be more specific for meter computation, and ideally we can also log the key which failed. same comments for `GetValue` above.

Reply:
i believe the meter error msg itself contains more info about which computation kind failed.  knowing the key in this case doesn't help too much since the user don't really have control over how cadence execute the code; the only thing the user can do is pay for more gas